src/shared_utilities/utils.jl

```diff
@@ -646,126 +594,117 @@ to check for NaNs based on the `nancheck_frequency`. The schedule is
 initialized with the `start_date` to ensure that it is first
 called at the correct time.
 """
-function NaNCheckCallback(
-    nancheck_frequency::Union{AbstractFloat, Dates.Period, ITime},
+function NaNCheckCallback(nancheck_frequency, start_date, dt; mask = nothing)
+
+    affect! = (integrator) -> call_count_nans_state(integrator.u; mask)
+
+    return FrequencyBasedCallback(nancheck_frequency, start_date, dt, affect!)
+end
+
+
+
+# TODO: Move to a more general callback system. For the time being, we use
+# the ClimaDiagnostics one because it is flexible and it supports calendar
+# dates.
+
+function FrequencyBasedCallback(
+    frequency::Union{AbstractFloat, Dates.Period, ITime},
     start_date,
-    dt;
-    mask = nothing,
+    dt,
+    affect!;
+    initialize = SciMLBase.INITIALIZE_DEFAULT,
+    save_positions = (true, true), # this preserves the old behavior, but I'm not sure if
+    # this is a good default
 )
-    # TODO: Move to a more general callback system. For the time being, we use
-    # the ClimaDiagnostics one because it is flexible and it supports calendar
-    # dates.
-
-    if nancheck_frequency isa AbstractFloat
-        # Assume it is in seconds, but go through Millisecond to support
-        # fractional seconds
-        nancheck_frequency_period = Dates.Millisecond(1000nancheck_frequency)
+    # Normalize floating point frequency to a Dates.Period
+    frequency_period =
+        frequency isa AbstractFloat ? Dates.Millisecond(1000 * frequency) :
+        frequency
+
+    dt_period = if typeof(dt) <: typeof(frequency_period) || isnothing(dt)
```

Comment:
Could you explain what happens if dt is nothing here? dt_period  = nothing, and then ?
and what start_date is used for (in a doc string?)